naga/src/front/wgsl/parse/mod.rs

```diff
@@ -2287,11 +2287,17 @@ impl Parser {
                                         let value = loop {
                                             let value = this.switch_value(lexer, ctx)?;
                                             if lexer.skip(Token::Separator(',')) {
-                                                if lexer.skip(Token::Separator(':')) {
+                                                // list of values ends with ':' or a compound statement
+                                                // indicated by an attribute or '{'
+                                                if matches!(
+                                                    lexer.peek().0,
+                                                    Token::Separator(':')
+                                                        | Token::Attribute
+                                                        | Token::Paren('{')
```

Comment:
I'd suggest making a helper `Token::is_start_of_compound_statement` for `matches!(token, Token::Attribute | Token::Paren('{'))`. It could be used here and on line 2124 where `fn statement` parses compound statements. This better encapsulates the sensitivity to a rule from elsewhere in the grammar and reduces the risk of bugs if the grammar changes in the future.

Reply:
That's a good idea! Adding the helper to `Token` didn't feel right to me, so I've added a free function instead.